sdk/src/azure/core/az_mqtt5_rpc.c

```diff
@@ -0,0 +1,191 @@
+#include <azure/core/internal/az_precondition_internal.h>
+#include <stdio.h>
+
+#if defined(TRANSPORT_MOSQUITTO)
+#include <mosquitto.h>
+#endif
+
+#include <azure/core/_az_cfg.h>
+
+AZ_NODISCARD bool az_span_topic_matches_sub(az_span sub, az_span topic)
+{
+  bool ret;
+#if defined(TRANSPORT_MOSQUITTO)
+  if (MOSQ_ERR_SUCCESS
+      != mosquitto_topic_matches_sub((char*)az_span_ptr(sub), (char*)az_span_ptr(topic), &ret))
+  {
+    ret = false;
+  }
+#else // TRANSPORT_MOSQUITTO
+  (void)sub;
+  (void)topic;
+  ret = false;
+#endif
+  return ret;
+}
+
+AZ_NODISCARD bool az_mqtt5_rpc_status_failed(az_mqtt5_rpc_status status)
+{
+  return (status < 200 || status >= 300);
+}
+
+AZ_NODISCARD az_result az_rpc_get_topic_from_format(
+    az_span model_id,
+    az_span executor_client_id,
+    az_span invoker_client_id,
+    az_span command_name,
+    az_span format,
+    az_span out_topic,
+    int32_t* out_topic_length)
+{
+  const az_span model_id_key = AZ_SPAN_FROM_STR("{serviceId}");
+  const az_span command_name_key = AZ_SPAN_FROM_STR("{name}");
+  const az_span executor_id_key = AZ_SPAN_FROM_STR("{executorId}");
+  const az_span invoker_id_key = AZ_SPAN_FROM_STR("{invokerId}");
+
+  int32_t format_size = az_span_size(format);
+
+  // Determine the length of the final topic
+  int32_t topic_length = format_size;
+  bool has_substitution = false;
+  int32_t max_temp_length = 0;
+  if (az_span_find(format, model_id_key) >= 0)
+  {
+    has_substitution = true;
+    _az_PRECONDITION_VALID_SPAN(model_id, 1, false);
+    topic_length += az_span_size(model_id);
+    topic_length -= az_span_size(model_id_key);
+    max_temp_length = topic_length > max_temp_length ? topic_length : max_temp_length;
+  }
+  if (az_span_find(format, command_name_key) >= 0)
+  {
+    has_substitution = true;
+    _az_PRECONDITION_VALID_SPAN(command_name, 1, false);
+    topic_length += az_span_size(command_name);
+    topic_length -= az_span_size(command_name_key);
+    max_temp_length = topic_length > max_temp_length ? topic_length : max_temp_length;
+  }
+  if (az_span_find(format, executor_id_key) >= 0)
+  {
+    has_substitution = true;
+    _az_PRECONDITION_VALID_SPAN(executor_client_id, 1, false);
+    topic_length += az_span_size(executor_client_id);
+    topic_length -= az_span_size(executor_id_key);
+    max_temp_length = topic_length > max_temp_length ? topic_length : max_temp_length;
+  }
+  if (az_span_find(format, invoker_id_key) >= 0)
+  {
+    has_substitution = true;
+    _az_PRECONDITION_VALID_SPAN(invoker_client_id, 1, false);
+    topic_length += az_span_size(invoker_client_id);
+    topic_length -= az_span_size(invoker_id_key);
+    max_temp_length = topic_length > max_temp_length ? topic_length : max_temp_length;
+  }
+
+  // Must be large enough to fit the entire format as items are substituted throughout the function
+  // even if that's larger than the output topic
+  _az_PRECONDITION_VALID_SPAN(out_topic, max_temp_length, true);
```

Comment:
So at this point `max_temp_length` will be equal to `topic_length`. Why not use `topic_length` instead then? and not declare `max_temp_length`

Reply:
Not necessarily - topic_length is the final topic length. If we create the topic in place in the out_topic span, then it needs to be able to have some keys and some substituted values. Ex. format is "type/{invokerId}/{executorId}/{name}/{serviceId}\0" and invokerId is "1", executorId is "2", etc, then the out_topic needs to be able to hold 
"type/{invokerId}/{executorId}/{name}/{serviceId}\0"  and then 
"type/1/{executorId}/{name}/{serviceId}\0"  and then 
"type/1/2/{name}/{serviceId}\0"   and then 
"type/1/2/3/{serviceId}\0" and then finally the topic_length span:
"type/1/2/3/4\0"

I might be able to get around this by creating another temporary az_span to have that longest topic in and then out_topic wouldn't need to be as large - I was having trouble getting that working the other day, but I might have some more ideas to try.